File: parentheses/balancer.rs
use std::collections::{HashMap, HashSet};

fn main() {
    // Test cases
    let test_cases = vec![
        "Hello (world)",
        "aaa(b]b(c(ddd)d]",
        "This is a [test] with {nested (parentheses)}",
        "Unbalanced ((example]",
        "CJK examples: 「こんにちは」「世界」",
        "Mixed types: (「Hello」)",
        "Invalid closing parenthesis: test)"
    ];

    println!("Testing parentheses balancer:");
    for test in test_cases {
        println!("Input: {}", test);
        println!("Result: {}", parentheses_balancer(test));
        println!("------------------------");
    }
}

fn parentheses_balancer(input: &str) -> String {
    // Define the pairs using a HashMap
    let pairs = create_pairs();
    
    // Create set of closing parentheses
    let values: HashSet<char> = pairs.values().cloned().collect();
    
    let mut stack = Vec::new();
    let mut result = String::new();

    for c in input.chars() {
        if pairs.contains_key(&c) {
            // If c is an opening parenthesis
            stack.push(c);
            result.push(c);
        } else if values.contains(&c) {
            // If c is a closing parenthesis
            if !stack.is_empty() && pairs.get(&stack[stack.len() - 1]) == Some(&c) {
                // Parentheses matches
                result.push(c);
                stack.pop();
            } else {
                // Parentheses does not match
                result.push(c);
            }
        } else {
            // If c isn't a parenthesis
            result.push(c);
        }
    }

    // Add closing parentheses for any remaining opening ones
    while !stack.is_empty() {
        let opening = stack.pop().unwrap();
        if let Some(closing) = pairs.get(&opening) {
            result.push(*closing);
        }
    }

    result
}

fn create_pairs() -> HashMap<char, char> {
    let mut pairs = HashMap::new();
    
    // Basic Latin parentheses
    pairs.insert('(', ')');
    pairs.insert('[', ']');
    pairs.insert('{', '}');
    pairs.insert('<', '>');  // Angle brackets
    
    // Guillemets (French quotes)
    pairs.insert('«', '»');
    pairs.insert('‹', '›');  // Single Guillemets
    
    // CJK parentheses (fullwidth forms)
    pairs.insert('（', '）'); // Fullwidth parenthesis
    pairs.insert('【', '】'); // Fullwidth square brackets
    pairs.insert('｛', '｝'); // Fullwidth curly brackets
    pairs.insert('〈', '〉'); // Fullwidth angle brackets
    pairs.insert('《', '》'); // Double angle brackets
    pairs.insert('「', '」'); // Corner brackets
    pairs.insert('『', '』'); // White corner brackets
    pairs.insert('〔', '〕');
    pairs.insert('〖', '〗');
    pairs.insert('〘', '〙');
    pairs.insert('〚', '〛');
    
    pairs
} 

File: applications/parentheses/balancer.cpp
#include <iostream>
#include <stack>
#include <unordered_map>
#include <unordered_set>
#include <string>
#include <sstream>
#include <vector>

// Using wide characters (wchar_t) to handle Unicode
using CharType = wchar_t;
using namespace std::string_literals;

// Define the pairs using an initializer list
const std::unordered_map<CharType, CharType> pairs = {
    {'(', ')'},  // Basic Latin parentheses
    {'[', ']'},
    {'{', '}'},
    {'<', '>'},  // Angle brackets
    {'«', '»'},  // Guillemets (French quotes)
    {'‹', '›'}, // Single Guillemets

    // CJK parentheses (fullwidth forms)
    {'（', '）'}, // Fullwidth parenthesis
    {'【', '】'}, // Fullwidth square brackets
    {'｛', '｝'}, // Fullwidth curly brackets
    {'〈', '〉'}, // Fullwidth angle brackets
    {'《', '》'}, // Double angle brackets
    {'「', '」'}, // Corner brackets
    {'『', '』'}, // White corner brackets
    {'〔', '〕'},
    {'〖', '〗'},
    {'〘', '〙'},
    {'〚', '〛'},
};

// Predefined set of values (closing parentheses)
const std::unordered_set<CharType> values = []() {
    std::unordered_set<CharType> vals;
    for (const auto& pair : pairs) {
        vals.insert(pair.second);
    }
    return vals;
}(); // Immediately invoked lambda to initialize the set



std::string parenthesesBalancer(const std::wstring& str) {
    std::stack<CharType> stack;
    std::wstringstream ss;

    for (CharType c : str) {
        if (pairs.contains(c)) {
            // If c is an opening parenthesis
            stack.push(c);
            ss << c;
        } else if (values.contains(c)) {
            // If c is a closing parenthesis
            if (!stack.empty() && pairs.at(stack.top()) == c) {
                // Parentheses matches
                ss << c;
                stack.pop();
            } else {
                // Parentheses does not match
                ss << c;
            }
        } else {
            // If c isn't a parenthesis
            ss << c;
        }
    }

    while (!stack.empty()) {
        CharType opening = stack.top();
        ss << pairs.at(opening);
        stack.pop();
    }

    std::wstring result = ss.str();
    return std::string(result.begin(), result.end());
}

int main() {
    // Test cases
    std::vector<std::wstring> testCases = {
        L"Hello (world)",
        L"aaa(b]b(c(ddd)d]",
        L"This is a [test] with {nested (parentheses)}",
        L"Unbalanced ((example]",
        L"CJK examples: 「こんにちは」「世界」",
        L"Mixed types: (「Hello」)",
        L"Invalid closing parenthesis: test)"
    };

    std::cout << "Testing parentheses balancer:\n";
    for (const auto& test : testCases) {
        std::wcout << L"Input: " << test << std::endl;
        std::cout << "Result: " << parenthesesBalancer(test) << std::endl;
        std::cout << "------------------------\n";
    }

    return 0;
}